src/downloads_counter.rs

```diff
@@ -0,0 +1,149 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// crates.io receives a lot of download requests, and we can't execute a write query to the
+/// database during each connection for performance reasons. To reduce the write load, this struct
+/// collects the pending updates from the current process and writes in batch.
+///
+/// To avoid locking the whole data structure behind a RwLock, which could potentially delay
+/// requests, this uses the dashmap crate. A DashMap has the same public API as an HashMap, but
+/// stores the items into `num_cpus()*4` individually locked shards. This approach reduces the
+/// likelyhood of a request encountering a locked shard.
+///
+/// Persisting the download counts in the database also takes advantage of the inner sharding of
+/// DashMaps: to avoid locking all the download requests at the same time each iteration only
+/// persists a single shard at the time.
+///
+/// The disadvantage of this approach is that download counts are stored in memory until they're
+/// persisted, so it's possible to lose some of them if the process exits ungracefully. While
+/// that's far from ideal, the advantage of batching database updates far outweights potentially
+/// losing some download counts.
+#[derive(Debug)]
+pub struct DownloadsCounter {
+    inner: DashMap<i32, AtomicUsize>,
+    shard_idx: AtomicUsize,
+    pending_count: AtomicUsize,
+}
+
+impl DownloadsCounter {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: DashMap::new(),
+            shard_idx: AtomicUsize::new(0),
+            pending_count: AtomicUsize::new(0),
+        }
+    }
+
+    pub(crate) fn increment(&self, version_id: i32) {
+        self.pending_count.fetch_add(1, Ordering::SeqCst);
+
+        if let Some(counter) = self.inner.get(&version_id) {
+            counter.value().fetch_add(1, Ordering::SeqCst);
+        } else {
+            self.inner
+                .entry(version_id)
+                .and_modify(|counter| {
+                    counter.fetch_add(1, Ordering::SeqCst);
+                })
+                .or_insert_with(|| AtomicUsize::new(1));
+        }
+    }
+
+    pub fn persist_all_shards(&self, app: &App) -> Result<(), Error> {
+        let conn = app.primary_database.get()?;
+
+        let mut counted_downloads = 0;
+        let mut counted_versions = 0;
+        let mut pending_downloads = 0;
+        for shard in self.inner.shards() {
+            let shard = std::mem::take(&mut *shard.write());
+            let stats = self.persist_shard(&conn, shard)?;
+
+            counted_downloads += stats.counted_downloads;
+            counted_versions += stats.counted_versions;
+            pending_downloads = stats.pending_downloads;
+        }
+
+        println!(
+            "download_counter all_shards counted_versions={} counted_downloads={} pending_downloads={}",
+            counted_versions,
+            counted_downloads,
+            pending_downloads,
+        );
```

Comment:
out of curiosity, is there a reason why we're using `println!()` everywhere instead of the `log` or `tracing` macros?

Reply:
I don't think there is really a reason, but I don't see value in switching to another library right now.